File: server/freen_phase.js
import express from 'express';
import cors from 'cors';
import fs from 'fs';
import { parse } from 'csv-parse/sync';

const app = express();
const PORT = 5000;

app.use(cors());
app.use(express.json());

// API to fetch categorized products
app.get('/api/products', (req, res) => {
    const csvData = fs.readFileSync('sample_dataset.csv', 'utf8');
    const records = parse(csvData, {
        columns: true,
        skip_empty_lines: true
    });

    // Convert dates to Date objects
    records.forEach(record => {
        record.manufacturing_date = new Date(record.manufacturing_date);
        record.expiry_date = new Date(record.expiry_date);
    });

    // Lists to store red, yellow, and green products
    const red_products = [];
    const yellow_products = [];
    const green_products = [];

    // Calculate the difference in months
    records.forEach(record => {
        const mfg_date = record.manufacturing_date;
        const exp_date = record.expiry_date;
        const difference_in_months = (exp_date.getFullYear() - mfg_date.getFullYear()) * 12 + exp_date.getMonth() - mfg_date.getMonth();
        if (difference_in_months <= 1) {
            record.tag = "red";
            red_products.push(record);
        } else if (difference_in_months <= 3) {
            record.tag = "yellow";
            yellow_products.push(record);
        } else {
            record.tag = "green";
            green_products.push(record);
        }
    });

    res.json({
        red: red_products,
        yellow: yellow_products,
        green: green_products
    });
});

// Start the server
app.listen(PORT, () => {
    console.log(`Server is running on http://localhost:${PORT}`);
});

File: server/freen_phase.py
import pandas as pd
from datetime import datetime
from tabulate import tabulate

# Read data from CSV file
df = pd.read_csv('sample_dataset.csv')

# Convert dates to datetime objects
df["manufacturing_date"] = pd.to_datetime(df["manufacturing_date"], format="%Y-%m-%d")
df["expiry_date"] = pd.to_datetime(df["expiry_date"], format="%Y-%m-%d")

# Lists to store red, yellow, and green products
red_products = []
yellow_products = []
green_products = []

# Calculate the difference in months
for index, row in df.iterrows():
    mfg_date = row["manufacturing_date"]
    exp_date = row["expiry_date"]
    difference_in_months = (exp_date.year - mfg_date.year) * 12 + exp_date.month - mfg_date.month
    if difference_in_months <= 1:
        row["tag"] = "red"
        red_products.append(row)
    elif difference_in_months <= 3:
        row["tag"] = "yellow"
        yellow_products.append(row)
    else:
        row["tag"] = "green"
        green_products.append(row)

# Convert lists to DataFrames
red_df = pd.DataFrame(red_products)
yellow_df = pd.DataFrame(yellow_products)
green_df = pd.DataFrame(green_products)

# Display data in table format
print("Red Products:")
print(tabulate(red_df, headers='keys', tablefmt='psql'))

print("\nYellow Products:")
print(tabulate(yellow_df, headers='keys', tablefmt='psql'))

print("\nGreen Products:")
print(tabulate(green_df, headers='keys', tablefmt='psql'))